### ISSUE TITLE
dual output (dict), one loss (dict) leads to border effect and error when saving weights

### ISSUE THREAD

**Issue by drauh** [keras-team/keras#16356]
### **Describe the problem**.

Dual model output (as dict), with only one loss (as dict, only matching one output) can lead to an error. This happened because the inputted dict to the compile method is modified by a side effect, modifying also the keras.Model.loss attribute. 

### **Standalone code to reproduce the issue**.

On Tensorflow 2.8, Python 3.9.7

**Inputs**
```
import keras

x = keras.layers.Input(shape=(1,))
model = keras.Model(
    inputs=x, outputs={"out1":x, "out2":x}
)
loss_dict = {"out1": "mse"}
model.compile(
    loss=loss_dict
)
model.fit(
    x=np.array([[1.0]]),
    y={"out1":np.array([[2.0]])},
    verbose=0
)
print(loss_dict)
print(model.loss)
model.save_weights("tmp")
```

**Outputs**
```
{'out1': 'mse', 'out2': None}
{'out1': 'mse', 'out2': None}
```

And saving weights raise this error :
```
ValueError: Unable to save the object {'out1': 'mse', 'out2': None} (a dictionary wrapper constructed automatically on attribute assignment). The wrapped dictionary was modified outside the wrapper (its final value was {'out1': 'mse', 'out2': None}, its value when a checkpoint dependency was added was None), which breaks restoration on object creation.

If you don't need this dictionary checkpointed, wrap it in a non-trackable object; it will be subsequently ignored.
```

### **Solution ?**

Adding a copy [here](https://github.com/keras-team/keras/blob/2db5acf3e3c5904b014cb409d3c514bef44f9640/keras/engine/compile_utils.py#L110) would resolve the problem in this particular situation.

**Comment by tilakrayal** [comment]
@gadagashwini ,
I was able to reproduce the issue in tf [v2.7](https://colab.research.google.com/gist/tilakrayal/41457c3ac0282c87f964c082b78f7495/untitled292.ipynb), v2.8 and nightly.Please find the gist [here](https://colab.research.google.com/gist/tilakrayal/762daf2ab47e08106cebc8a9ff48a0f2/untitled293.ipynb).

**Comment by gadagashwini** [comment]
@drauh,
Hi, Thanks for reporting this issue.
I modified your code in model definition. See the working sample code snippet below.

```
import tensorflow as tf
import numpy as np

# network that maps 1 input to 2 separate outputs
x = tf.keras.layers.Input(shape=(1,), dtype=tf.float32)
x2 = tf.keras.layers.Dense(2)(x)
y, z = tf.keras.layers.Lambda(
    lambda x2: tf.unstack(x2, axis=-1), name='out')(x2)

model = tf.keras.models.Model(inputs=x, outputs=dict(y=y, z=z))

print(model.outputs)
print(model.output_names)

model.compile('SGD', loss={'x': 'mse', 'y': 'logcosh'})
model.fit(
    x=np.array([[1.0]]),
    y =dict(y=np.array([[2.0]]), z=np.array([[2.0]])),
    #y={y:np.array([[2.0]]), z:np.array([[2.0]])},
    verbose=0
)
```
**Output**
```
[<KerasTensor: shape=(None,) dtype=float32 (created by layer 'out')>, <KerasTensor: shape=(None,) dtype=float32 (created by layer 'out')>]
['out', 'out_1']
```

**Comment by drauh** [comment]
In my use case, I need the model to have two outputs but only one loss for only one of the outputs. And it seemed to me that this is allowed. I'm working on #16360 to resolve this problem.

**Comment by drauh** [comment]
Solved by #16360
